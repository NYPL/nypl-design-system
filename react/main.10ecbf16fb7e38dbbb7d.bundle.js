(window.webpackJsonp=window.webpackJsonp||[]).push([[0],{1292:function(n,o,c){"use strict";c.r(o);c(1293),c(1323),c(1335),c(1442),c(1485)},1448:function(n,o){},1449:function(n,o){},1450:function(n,o){},1451:function(n,o){},501:function(n,o,c){c(502),c(645),n.exports=c(1292)},564:function(n,o){}},[[501,1,2]]]);