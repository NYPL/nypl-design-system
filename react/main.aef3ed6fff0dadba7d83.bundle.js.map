{"version":3,"file":"main.aef3ed6fff0dadba7d83.bundle.js","sources":["webpack:///main.aef3ed6fff0dadba7d83.bundle.js"],"mappings":"AAAA","sourceRoot":""}