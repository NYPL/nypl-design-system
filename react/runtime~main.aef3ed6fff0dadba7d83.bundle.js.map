{"version":3,"file":"runtime~main.aef3ed6fff0dadba7d83.bundle.js","sources":["webpack:///runtime~main.aef3ed6fff0dadba7d83.bundle.js"],"mappings":"AAAA","sourceRoot":""}