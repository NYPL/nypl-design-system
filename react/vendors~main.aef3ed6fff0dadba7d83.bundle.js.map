{"version":3,"file":"vendors~main.aef3ed6fff0dadba7d83.bundle.js","sources":["webpack:///vendors~main.aef3ed6fff0dadba7d83.bundle.js"],"mappings":";AAAA","sourceRoot":""}