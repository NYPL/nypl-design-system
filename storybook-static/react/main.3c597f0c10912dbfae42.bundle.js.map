{"version":3,"file":"main.3c597f0c10912dbfae42.bundle.js","sources":["webpack:///main.3c597f0c10912dbfae42.bundle.js"],"mappings":"AAAA","sourceRoot":""}