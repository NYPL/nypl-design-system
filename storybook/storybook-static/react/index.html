<!doctype html><html lang="en"><head><meta charset="utf-8"><title>Storybook</title><meta name="viewport" content="width=device-width,initial-scale=1"><style>html, body {
    overflow: hidden;
    height: 100%;
    width: 100%;
    margin: 0;
    padding: 0;
  }</style><script>/* globals window */
  /* eslint-disable no-underscore-dangle */
  try {
    if (window.top !== window) {
      window.__REACT_DEVTOOLS_GLOBAL_HOOK__ = window.top.__REACT_DEVTOOLS_GLOBAL_HOOK__;
    }
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn('unable to connect to top frame for connecting dev tools');
  }</script><style>#root[hidden],
      #docs-root[hidden] {
        display: none !important;
      }</style></head><body><div id="root"></div><div id="docs-root"></div><script>window['DOCS_MODE'] = false;</script><script src="runtime~main.00bd9d033d7a0b972fc9.bundle.js"></script><script src="vendors~main.d2b6aafef1a259d165c0.bundle.js"></script><script src="main.3bf71672a6dd35499004.bundle.js"></script></body></html>