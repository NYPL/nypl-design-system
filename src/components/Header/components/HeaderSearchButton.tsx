import FocusLock from "@chakra-ui/focus-lock";
import { Box, chakra, useStyleConfig } from "@chakra-ui/react";
import React, { useState, useRef } from "react";

import Button from "../../Button/Button";
import HeaderSearchForm from "./HeaderSearchForm";
import Icon from "../../Icons/Icon";
import useCloseDropDown from "../../../hooks/useCloseDropDown";
import gaUtils from "../utils/googleAnalyticsUtils";

export interface HeaderSearchButtonProps {
  isMobile?: boolean;
}

/**
 * This is the button that will render the search form when it is clicked and
 * keep focus trapped within the menu.
 */
const HeaderSearchButton = chakra(
  ({ isMobile = false }: HeaderSearchButtonProps) => {
    const [isOpen, setIsOpen] = useState<boolean>(false);
    const styles = useStyleConfig("HeaderSearchButton", { isOpen });
    const buttonText = isMobile ? null : isOpen ? "Close" : "Search";
    const labelText = isOpen ? "Close Search" : "Open Search";
    const ref = useRef<HTMLDivElement>(null);
    const gaAction = isMobile ? "Click" : "Search";
    const gaLabel = isMobile
      ? "Mobile clickSearch"
      : isOpen
      ? "Close Menu"
      : "Open Menu";

    useCloseDropDown(setIsOpen, ref);

    return (
      <Box ref={ref}>
        <FocusLock isDisabled={!isOpen}>
          <Button
            aria-haspopup="true"
            aria-label={labelText}
            aria-expanded={isOpen ? true : null}
            buttonType="text"
            id="searchButton"
            onClick={() => {
              gaUtils.trackEvent(gaAction, gaLabel);
              setIsOpen(!isOpen);
            }}
            __css={styles}
          >
            <span>
              {buttonText}
              <Icon
                align={isMobile ? "none" : "right"}
                name={isOpen ? "close" : "search"}
                size={isMobile ? "large" : isOpen ? "medium" : "small"}
                title={labelText}
              />
            </span>
          </Button>
          {isOpen && <HeaderSearchForm isMobile={isMobile} />}
        </FocusLock>
      </Box>
    );
  }
);

export default HeaderSearchButton;
