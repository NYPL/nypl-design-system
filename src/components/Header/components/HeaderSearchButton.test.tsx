import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { axe } from "jest-axe";
import * as React from "react";
import ReactGa from "react-ga";
import renderer from "react-test-renderer";

import HeaderSearchButton from "./HeaderSearchButton";
import gaUtils from "../utils/googleAnalyticsUtils";

gaUtils.initialize({ testMode: true }, false);

describe("HeaderSearchButton Accessibility", () => {
  it("passes axe accessibility test", async () => {
    const { container } = render(<HeaderSearchButton />);
    expect(await axe(container)).toHaveNoViolations();
  });

  it("passes axe accessibility test for mobile", async () => {
    const { container } = render(<HeaderSearchButton isMobile />);
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("HeaderSearchButton", () => {
  describe("Desktop", () => {
    beforeEach(() => {
      render(<HeaderSearchButton />);
    });

    it("renders a Search button", () => {
      const searchBtn = screen.getByRole("button");

      expect(searchBtn).toBeInTheDocument();
      expect(searchBtn).toHaveTextContent(/search/i);
    });

    it("renders a Close button when clicked", () => {
      const searchBtn = screen.getByRole("button");

      expect(searchBtn).toHaveTextContent(/search/i);
      expect(screen.getByTitle(/Open Search/i)).toBeInTheDocument();
      expect(screen.queryByTitle(/Close Search/i)).not.toBeInTheDocument();

      userEvent.click(searchBtn);

      expect(searchBtn).toHaveTextContent(/close/i);
      expect(screen.queryByTitle(/Open Search/i)).not.toBeInTheDocument();
      expect(screen.getByTitle(/Close Search/i)).toBeInTheDocument();
      // The first call is the initializing call.
      // The second call is what we want.
      expect(ReactGa.testModeAPI.calls[1]).toEqual([
        "send",
        {
          eventCategory: "Global Header",
          eventAction: "Search",
          eventLabel: "Open Menu",
          hitType: "event",
        },
      ]);
    });

    it("renders the search form when clicked", () => {
      const searchBtn = screen.getByRole("button");

      expect(screen.queryByRole("form")).not.toBeInTheDocument();

      userEvent.click(searchBtn);

      expect(screen.getByRole("form")).toBeInTheDocument();
    });
  });

  describe("Mobile", () => {
    beforeEach(() => {
      render(<HeaderSearchButton isMobile />);
    });

    it("renders a Search button", () => {
      const searchBtn = screen.getByRole("button");

      // On mobile, the text is not displayed but is instead
      // added as an aria-label.
      expect(searchBtn).not.toHaveTextContent(/search/i);
      expect(searchBtn).toHaveAttribute("aria-label", "Open Search");
    });

    it("renders a Close button when clicked", () => {
      const searchBtn = screen.getByRole("button");

      expect(searchBtn).toHaveAttribute("aria-label", "Open Search");

      userEvent.click(searchBtn);

      expect(searchBtn).toHaveAttribute("aria-label", "Close Search");
      // The third call is the click event for this test.
      expect(ReactGa.testModeAPI.calls[3]).toEqual([
        "send",
        {
          eventCategory: "Global Header",
          eventAction: "Click",
          eventLabel: "Mobile clickSearch",
          hitType: "event",
        },
      ]);
    });

    it("renders the search form when clicked", () => {
      const searchBtn = screen.getByRole("button");

      expect(screen.queryByRole("form")).not.toBeInTheDocument();

      userEvent.click(searchBtn);

      expect(screen.getByRole("form")).toBeInTheDocument();
    });
  });

  it("renders the UI snapshot correctly", () => {
    const headersearchButton = renderer.create(<HeaderSearchButton />).toJSON();
    const headersearchButtonMobile = renderer
      .create(<HeaderSearchButton isMobile />)
      .toJSON();

    expect(headersearchButton).toMatchSnapshot();
    expect(headersearchButtonMobile).toMatchSnapshot();
  });
});
