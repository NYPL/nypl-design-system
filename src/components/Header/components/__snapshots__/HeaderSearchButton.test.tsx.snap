// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderSearchButton renders the UI snapshot correctly 1`] = `
<div
  className="css-0"
>
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={-1}
  />
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={-1}
  />
  <div
    data-focus-lock-disabled="disabled"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <button
      aria-expanded={null}
      aria-haspopup="true"
      aria-label="Open Search"
      className="chakra-button css-1xdhyk6"
      id="searchButton"
      onClick={[Function]}
      type="button"
    >
      <span>
        Search
        <svg
          aria-hidden={true}
          className="chakra-icon css-1grhd2q"
          focusable={false}
          role="img"
          title="Open Search"
          viewBox="0 0 24 24"
        >
          <g
            stroke="currentColor"
            strokeWidth="1.5"
          >
            <path
              d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
              fill="none"
              strokeLinecap="round"
            />
            <path
              d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
              fill="currentColor"
              strokeLinecap="round"
            />
            <circle
              cx="12"
              cy="12"
              fill="none"
              r="11.25"
              strokeMiterlimit="10"
            />
          </g>
        </svg>
      </span>
    </button>
  </div>
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={-1}
  />
</div>
`;

exports[`HeaderSearchButton renders the UI snapshot correctly 2`] = `
<div
  className="css-0"
>
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={-1}
  />
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={-1}
  />
  <div
    data-focus-lock-disabled="disabled"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <button
      aria-expanded={null}
      aria-haspopup="true"
      aria-label="Open Search"
      className="chakra-button css-1xdhyk6"
      id="searchButton"
      onClick={[Function]}
      type="button"
    >
      <span>
        <svg
          aria-hidden={true}
          className="chakra-icon css-1grhd2q"
          focusable={false}
          role="img"
          title="Open Search"
          viewBox="0 0 24 24"
        >
          <g
            stroke="currentColor"
            strokeWidth="1.5"
          >
            <path
              d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
              fill="none"
              strokeLinecap="round"
            />
            <path
              d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
              fill="currentColor"
              strokeLinecap="round"
            />
            <circle
              cx="12"
              cy="12"
              fill="none"
              r="11.25"
              strokeMiterlimit="10"
            />
          </g>
        </svg>
      </span>
    </button>
  </div>
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={-1}
  />
</div>
`;
